3scale/porta
# [Threescale 7522] Fix CDN urls for fonts assets

**What this PR does / why we need it**:

In the STG-SaaS porta instance, Setting the `RAILS_ASSET_HOST` variable makes all resources to be loaded from the CDN, except the fonts.

**Which issue(s) this PR fixes** 

[THREESCALE-7522](https://issues.redhat.com/browse/THREESCALE-7522)

**Verification steps** 

1. Load any screen with the development tools open
2. In the network tab, check that all resources (Images, Fonts, Js, etc) are loaded from the CDN
3. Change the value of `RAILS_ASSET_HOST` and restart the server
4. In the network tab, check that all resources (Images, Fonts, Js, etc) are loaded from the new CDN

**Special notes for your reviewer**:

### Where the fonts urls come from:

- The fonts are originally declared in [Patternfly](https://github.com/patternfly/patternfly/blob/8aaa34bb40afb6b48ca35d0e5cb27ec8384a63ae/src/patternfly/base/_fonts.scss)
- Instead of simply transpiling the above file to a CSS, Webpack uses javascript to dynamically generate a CSS code equivalent to that, but in runtime. This design allows webpack to replace the CDN url without the need to recompile the assets again, which only happens once during the deploy/image build.
- When no `__webpack_public_path__` is given in runtime, the CSS stubs are generated using the URL at `settings.yml_asset_host` during compilation time. Thanks to webpacker, I guess.

### The problem:

- In the STG-SaaS server, the assets are compiled during the image build, having `settings.yml_asset_host == null` and thus setting relative urls for fonts by default.
- In runtime, the `__webpack_public_path__` variable is either empty or not available, then the JS generates a CSS code pointing to fonts using realtive urls, and thus being loaded from the server.
- However, the urls are correct in the SaaS server, that's because the CDN url is [hardcoded](https://github.com/3scale/deploy/blob/a0ecdb97ef327b140a6fc341c4528d91d70cb3a5/system/config/deploy/production.rb#L122) and available during assets compilation time

### Implemented solution:

I used a simplification of this [workaround](https://github.com/rails/webpacker/issues/2269#issuecomment-673398972), which:
1. Assumes the assets where compiled without any value at `settings.yml_asset_host`, thus generating relative paths.
2. Tricking webpack to prepend the CDN url in runtime.
3. Assuring the trick is not executed when a CDN is provided during assets compilation, like it SaaS deployment does.

**Observations:**
- Probably because of [this](https://github.com/rails/webpacker/issues/1080#issuecomment-351896551), I wasn't able to make webpack honor the value of `__webpack_public_path__`. Then I used the variable `window.rails_asset_host` instead, which allows us to simplify the solution and set the the value only once at `provider.html.slim`.
- The value of `window.rails_asset_host` must be set at the entry point of the app, before including any other JS. I placed it at `app/views/layouts/provider.html.slim`, not sure if it's the right place.
- I updated Webpacker to 5.4.3. This doesn't update webpack, only the Rails wrapper gem. I did it because there are no breaking changes [from 4.3.0 to 5.4.3](https://github.com/rails/webpacker/blob/master/CHANGELOG.md) and I found a [comment](https://github.com/rails/webpacker/issues/1080#issuecomment-350761506) complaining about how webpacker sets `WEBPACKER_ASSET_HOST`, which I use to determine whether the user is compiling the assets with a default URL. I think it could probably work without the update, so if the update causes problem to any of you, I could try to remove it.
- I also updated file-loader to 6.2.0, this is something I need because our old version doesn't support the `postTransformPublicPath` option, that I need. Again, if this causes problems, I could probably use an older version, as long as it provides the mentioned option.
